apps/req.c

```diff
@@ -1708,14 +1705,25 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
         && do_pkey_ctx_init(pkctx, sigopts);
 }
 
-int do_X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md,
+/* Ensure RFC 5280 compliance and then sign the certificate info */
+int do_X509_sign(X509 *cert, EVP_PKEY *pkey, const EVP_MD *md,
                  STACK_OF(OPENSSL_STRING) *sigopts)
 {
-    int rv = 0;
+    const STACK_OF(X509_EXTENSION) *exts = X509_get0_extensions(cert);
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
+    int rv = 0;
 
-    if (do_sign_init(mctx, pkey, md, sigopts) > 0)
-        rv = (X509_sign_ctx(x, mctx) > 0);
+    if (sk_X509_EXTENSION_num(exts /* may be NULL */) > 0) {
+        /* Prevent X509_V_ERR_EXTENSIONS_REQUIRE_VERSION_3 */
+        if (!X509_set_version(cert, 2)) /* Make sure cert is X509 v3 */
+            goto end;
+
+        /* TODO further measures for ensuring RFC 5280 compliance */
```

Comment:
I am not quite sure what do you mean by this. I do not think we should disallow making RFC 5280 non-compliant certificates. If you mean adding the SKID and AKID, then OK. Otherwise anything else should be probably just a warning if anything.

Reply:
Yeah, the addition of SKID and AKID (unless explicitly disabled) is what I meant primarily.
Likely there are more such things.

Generally, I'd say those apps should produce compliant certs by default.
Of course, at library level there is full freedom anyway to do weird things :wink:

Reply:
I've meanwhile re-phrased this to:
```
        /* any further measures for ensuring default RFC 5280 compliance? */
```